### PR TITLE
New Adapter: SA Lunamedia 

### PR DESCRIPTION
**Pull request by lunamedia**

### CODE REVIEW

**Comment by SyntaxNode** [comment]
How is this different from the existing LunaMedia adapter?

**Comment by lunamedia** [comment]
@SyntaxNode 
We have created a new adapter, will test it with several publishers and if the test is successful, we deprecate the old one.

File: static/bidder-params/sa_lunamedia.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        "type": "string",
+        "description": "network or placement key"
+      },
+      "type": {
+        "type": "string",
+        "enum": ["network", "publisher"]
+      }
+    },
+    "required": ["key"]
+  }
```
Comment (VeronikaSolovei9) [inline static/bidder-params/sa_lunamedia.json:17]: @SyntaxNode there is no `openrtb_ext/imp_{bidder}.go` file and no `ImpExt{Bidder}` despite there are some bidder specific parameters are expected (key and type). Is it ok not to have this file?

File: adapters/sa_lunamedia/salunamedia.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+	if len(bidResp.SeatBid) == 0 {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Empty SeatBid"),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	bids := bidResp.SeatBid[0].Bid
```
Comment (VeronikaSolovei9) [inline adapters/sa_lunamedia/salunamedia.go:98]: Should it be a loop over all SeatBids instead?

File: adapters/sa_lunamedia/salunamedia.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+	if len(bids) == 0 {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Empty SeatBid.Bids"),
+		}}
+	}
+
+	bid := bids[0]
```
Comment (VeronikaSolovei9) [inline adapters/sa_lunamedia/salunamedia.go:106]: Should it be a loop over all Bids instead?

File: adapters/sa_lunamedia/salunamedia.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+		return nil, []error{err}
+	}
+
+	if len(bidResp.SeatBid) == 0 {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Empty SeatBid"),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
```
Comment (VeronikaSolovei9) [inline adapters/sa_lunamedia/salunamedia.go:96]: Does this mean response always has only one bid even if impression number was more than 1?

File: adapters/sa_lunamedia/salunamedia.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+	if bidderRawResponse.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Bad Request. %s", string(bidderRawResponse.Body)),
+		}}
+	}
+
+	if bidderRawResponse.StatusCode == http.StatusServiceUnavailable {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Bidder unavailable."),
```
Comment (SyntaxNode) [inline adapters/sa_lunamedia/salunamedia.go:74]: There are no arguments passed to `Sprintf`. Please just include the string literal. There are a few other instances of this in this method.

**Review by SyntaxNode — COMMENTED** [review]
Looks good. I have a few minor comments. Please open a docs PR per https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#user-documentation.

**Review by AlexBVolcy — CHANGES_REQUESTED** [review]
I'll wait on an approval until after other comments are addressed!

**Review by AlexBVolcy — COMMENTED** [review]
Waiting on other comments to be addressed before approval!

File: adapters/sa_lunamedia/salunamedia.go (hunk 2 above)
Comment (lunamedia) [reply adapters/sa_lunamedia/salunamedia.go:98]: Our server always gives only one bid.

File: adapters/sa_lunamedia/salunamedia.go (hunk 3 above)
Comment (lunamedia) [reply adapters/sa_lunamedia/salunamedia.go:106]: Our server does not return a multi bid.

File: adapters/sa_lunamedia/salunamedia.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	if bidderRawResponse.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Bad Request. %s", string(bidderRawResponse.Body)),
+		}}
+	}
+
+	if bidderRawResponse.StatusCode == http.StatusServiceUnavailable {
+		return nil, []error{&errortypes.BadInput{
```
[the remark this answers is not in the document]
Comment (lunamedia) [reply adapters/sa_lunamedia/salunamedia.go:73]: We have added a text that the publisher would contact our support.

File: static/bidder-params/sa_lunamedia.json (hunk 1 above)
Comment (lunamedia) [reply static/bidder-params/sa_lunamedia.json:17]: We have added ExtImpSaLunamedia. But we didn't use it in the adapter. Do we need to use it for validation? I thought that such a check is already happening at the prebid-server level ?

File: adapters/sa_lunamedia/params_test.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+var invalidParams = []string{
+	``,
+	`null`,
+	`true`,
+	`5`,
+	`4.2`,
+	`[]`,
+	`{}`,
+	`{ "anyparam": "anyvalue" }`,
```
[the remark this answers is not in the document]
Comment (lunamedia) [reply adapters/sa_lunamedia/params_test.go:36]: Added new test cases.

File: adapters/sa_lunamedia/salunamedia.go (hunk 5 above)
Comment (lunamedia) [reply adapters/sa_lunamedia/salunamedia.go]: Changed

**Comment by lunamedia** [comment]
@AlexBVolcy @SyntaxNode @VeronikaSolovei9 
We have made changes and responded to comments.

File: static/bidder-params/sa_lunamedia.json (hunk 1 above)
Comment (VeronikaSolovei9) [reply static/bidder-params/sa_lunamedia.json:17]: Yes, validation happens on Prebid server level based on sa_lunamedia.json file. You don't need to explicitly validate it in adapter code.

**Comment by SyntaxNode** [comment]
We'll re-review this today, but before we can consider merging the PR in we need you to first open a docs PR.

**Review by VeronikaSolovei9 — COMMENTED** [review]
LGTM

**Review by AlexBVolcy — APPROVED** [review]
LGTM

**Comment by SyntaxNode** [comment]
The PR looks good. We'll be able to merge it once the docs PR is opened and linked here. Thank you.

**Comment by lunamedia** [comment]
@SyntaxNode 
https://github.com/prebid/prebid.github.io/pull/3082

**Review by VeronikaSolovei9 — APPROVED** [review]

**Pull request merged**